atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/LocalWriteBufferTest.java

```diff
@@ -0,0 +1,208 @@
+        assertThat(buffer.getChangeMetadataForWritesToTable(TABLE2))
+                .containsExactlyInAnyOrderEntriesOf(ImmutableMap.of(CELL_1, METADATA_1, CELL_2, METADATA_2));
+    }
+
+    @Test
+    public void cannotPutMetadataWithoutAssociatedWrite() {
+        assertThatLoggableExceptionThrownBy(() ->
+                        buffer.putLocalWritesAndMetadata(TABLE, ImmutableMap.of(), ImmutableMap.of(CELL_1, METADATA_1)))
+                .isInstanceOf(SafeIllegalStateException.class)
+                .hasLogMessage("Every metadata we put must be associated with a write")
+                .hasExactlyArgs(
+                        UnsafeArg.of("tableRef", TABLE),
+                        UnsafeArg.of("cellsWithOnlyMetadata", ImmutableSet.of(CELL_1)));
+    }
+
+    /**
+     * This test would fail for a previous synchronization model for {@link LocalWriteBuffer#putLocalWritesAndMetadata},
+     * where we used {@link ConcurrentSkipListMap#compute} in the following way:
+     *
+     * <pre> {@code
+     *  writes.compute(cell, (k, oldVal) -> {
+     *                 if (hasMetadata) {
+     *                     metadataForWrites.put(cell, metadata.get(cell));
+     *                 } else {
+     *                     metadataForWrites.remove(cell);
+     *                 }
+     *                 return val;
+     *             });
+     * }</pre>
+     * <p>
+     * {@link ConcurrentSkipListMap#compute} uses {@link AtomicReference#compareAndSet} to test whether another thread
+     * has set the value in the meantime. It uses the old value for the key read before calling
+     * {@link ConcurrentSkipListMap#compute} to decide whether it should retry, including re-computation of the value.
+     * <p>
+     * Let Cell C hold value A and no metadata and the following sequence involving two threads occurs:
+     * <pre> {@code
+     * |                   T1                   |                    T2                   |
+     * |:--------------------------------------:|:---------------------------------------:|
+     * | read A as old value                    |                                         |
+     * | putMetadata(C, X)                      |                                         |
+     * |                                        | read A as old value                     |
+     * |                                        | putMetadata(C, Y)                       |
+     * |                                        | compareAndSet(expected: A, newValue: A) |
+     * | compareAndSet(expected:A, newValue: B) |                                         |
+     * }</pre>
+     * Because T2 did not modify the value of C, the {@link AtomicReference#compareAndSet} in T1 succeeds.
+     * At the end of this sequence, C has value B and metadata Y, which violates the atomicity guarantee.
+     */
+    @Test
+    public void writingValueAndMetadataIsAtomicWhenOverwritingWithSameValue() throws InterruptedException {
```

Comment:
I removed `SnapshotTransactionConcurrencyTest` in favor of this (would have been a lot of effort for a nearly duplicate test)

Reply:
also the beauty of splitting this out!